capsules/src/net/ipv6/ipv6.rs

```diff
@@ -0,0 +1,324 @@
+            TransportHeader::ICMP(icmp_header) => icmp_header.encode(buf, offset).done().unwrap(),
+            _ => {
+                unimplemented!();
+            }
+        };
+        let payload_length = self.get_payload_length();
+        let offset = enc_consume!(buf, offset; encode_bytes, &self.payload[..payload_length]);
+        stream_done!(offset, offset)
+    }
+
+    fn get_payload_length(&self) -> usize {
+        match self.header {
+            TransportHeader::UDP(udp_header) => {
+                udp_header.get_len() as usize - udp_header.get_hdr_size()
+            }
+            TransportHeader::ICMP(icmp_header) => {
+                icmp_header.get_len() as usize - icmp_header.get_hdr_size()
+            }
+            _ => {
+                unimplemented!();
+            }
+        }
+    }
+}
+
+pub struct IP6Packet<'a> {
+    pub header: IP6Header,
+    pub payload: IPPayload<'a>,
+}
+
+// Note: We want to have the IP6Header struct implement these methods,
+// as there are cases where we want to allocate/modify the IP6Header without
+// allocating/modifying the entire IP6Packet
+impl<'a> IP6Packet<'a> {
+    // Sets fields to appropriate defaults
+
+    pub fn new(pyld: IPPayload<'a>) -> IP6Packet<'a> {
+        IP6Packet {
+            header: IP6Header::default(),
+            payload: pyld,
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.header = IP6Header::default();
+    }
+
+    pub fn get_total_len(&self) -> u16 {
+        40 + self.header.get_payload_len()
+    }
+
+    pub fn get_payload(&self) -> &[u8] {
+        self.payload.payload
+    }
+
+    pub fn get_total_hdr_size(&self) -> usize {
+        let transport_hdr_size = match self.payload.header {
+            TransportHeader::UDP(udp_hdr) => udp_hdr.get_hdr_size(),
+            TransportHeader::ICMP(icmp_header) => icmp_header.get_hdr_size(),
+            _ => unimplemented!(),
+        };
+        40 + transport_hdr_size
+    }
+
+    pub fn set_transport_checksum(&mut self) {
+        //Looks at internal buffer assuming
+        // it contains a valid IP packet, checks the payload type. If the payload
+        // type requires a cksum calculation, this function calculates the
+        // psuedoheader cksum and calls the appropriate transport packet function
+        // using this pseudoheader cksum to set the transport packet cksum
+
+        match self.payload.header {
+            TransportHeader::UDP(ref mut udp_header) => {
+                let cksum = compute_udp_checksum(
+                    &self.header,
+                    &udp_header,
+                    udp_header.get_len(),
+                    self.payload.payload,
+                );
+                udp_header.set_cksum(cksum);
+            }
+            TransportHeader::ICMP(ref mut icmp_header) => {
+                let cksum = compute_icmp_checksum(&self.header, &icmp_header, self.payload.payload);
+                icmp_header.set_cksum(cksum);
```

Comment:
Wouldn't it make sense to collapse `compute_*_checksum` and `set_cksum` into methods of a trait that each transport header type implements? It seems like it makes more sense for the module implementing the transport layer to implement the checksum computation